it/uk/gov/hmrc/agentsubscriptionfrontend/controllers/SubscriptionControllerISpec.scala

```diff
@@ -84,90 +87,249 @@ class SubscriptionControllerISpec extends BaseISpec with SessionDataMissingSpec
+      sessionStoreService.currentSession.knownFactsResult = Some(myAgencyKnownFactsResult)
+    }
 
-      val result =
-        await(controller.showSubscriptionComplete(request.withFlash("arn" -> "AARN0000001", "agencyName" -> "My Agency")))
+    def resultOf(request: Request[AnyContent]) = await(controller.submitLinkAccount(request))
 
-      status(result) shouldBe 200
-      checkHtmlResultWithBodyText(result, "You must save this number for your agency's records.")
-      checkHtmlResultWithBodyText(result, "AARN-000-0001")
+    behave like anAgentAffinityGroupOnlyEndpoint(resultOf)
+
+    "choice is Yes" should {
+      "redirect to /subscription-complete" in new RequestWithSessionDetails(autoMappingFormValue = "yes") {
+        MappingStubs.givenMappingUpdateToPostSubscription(utr)
+        resultOf(request).header.headers(LOCATION) shouldBe routes.SubscriptionController.showSubscriptionComplete().url
+      }
+
+      "keep the ARN in the session" in new RequestWithSessionDetails(autoMappingFormValue = "yes") {
+        MappingStubs.givenMappingUpdateToPostSubscription(utr)
+        resultOf(request).session.get("arn") shouldBe Some("AARN0000001")
+      }
+
+      "update the pre-subscription mappings with the ARN" in new RequestWithSessionDetails(autoMappingFormValue = "yes") {
+        MappingStubs.givenMappingUpdateToPostSubscription(utr)
+        val result = resultOf(request)
+        MappingStubs.verifyMappingUpdateToPostSubscriptionCalled(utr)
+      }
+    }
+
+    "choice is No" should {
+      "redirect to /subscription-complete" in new RequestWithSessionDetails(autoMappingFormValue = "no") {
+        MappingStubs.givenMappingDeletePreSubscription(utr)
+        resultOf(request).header.headers(LOCATION) shouldBe routes.SubscriptionController.showSubscriptionComplete().url
+      }
+
+      "keep the ARN in the session" in new RequestWithSessionDetails(autoMappingFormValue = "no") {
+        MappingStubs.givenMappingDeletePreSubscription(utr)
+        resultOf(request).session.get("arn") shouldBe Some("AARN0000001")
+      }
+
+      "delete the pre-subscription mappings" in new RequestWithSessionDetails(autoMappingFormValue = "no") {
+        MappingStubs.givenMappingDeletePreSubscription(utr)
+        val result = resultOf(request)
+        MappingStubs.verifyMappingDeletePreSubscriptionCalled(utr)
+      }
     }
 
-    "redirect to session missing page if there is nothing in the flash scope" in {
+    "choice is missing" should {
+      "return 200 and redisplay the /link-account page with an error message for missing choice" in {
+        val request = authenticatedAs(subscribingAgentEnrolledForHMRCASAGENT)
+          .withSession("arn" -> "AARN0000001")
+
+        resultOf(request) should containMessages("linkAccount.title", "error.no-radio-selected")
+      }
+    }
+
+    "form value is invalid" should {
+      "result in a BadRequest" in new RequestWithSessionDetails(autoMappingFormValue = "somethingInvalid") {
```

Comment:
I think it will return a match error which may result in internal server error other than bad request. We could throw a bad-request-exception for invalid values of that form.

Reply:
Done